### ISSUE TITLE
New received messages not delivered after a "Have not received a heartbeat in 900 secondes"  

### ISSUE THREAD

**Issue by rejoc** [gmr/rabbitpy#74]
rabbitpy 0.26.2

I have a simple program that reads messages on a queue and store them in a DB : 
```
        with rabbitpy.Connection(p.url) as connection:
            log.info('Connected to rabbitmq')
            with connection.channel() as channel:
                log.info('Channel opened')
                queue = rabbitpy.Queue(channel, rabbit_queue)
                for message in queue.consume_message():
                    log.info('Message: %r, %s', message, message.routing_key)
                    value = message.body
                    if isinstance(value, str):
                        value = value.decode('utf-8')
                    if message.properties.get('content_type', '').lower() == 'application/json':
                        value = from_json(value)
                    statistic_collection.save(value)
                    message.ack()

```

Sometime, there is a heartbeat miss. No error is raised up to the program so it is still in the ```for message in``` loop. But the new messages are not yield to the loop. 

The log is here: 

```
2015-05-18 14:21:08,651 - rabbitpy.base - DEBUG - Connection setting state to 'Opening'
2015-05-18 14:21:08,652 - rabbitpy.events - DEBUG - Waiting for 1 seconds on event: Socket Connected
2015-05-18 14:21:08,652 - rabbitpy.base - DEBUG - IO setting state to 'Opening'
2015-05-18 14:21:09,652 - rabbitpy.events - DEBUG - Waiting for 1 seconds on event: Socket Connected
2015-05-18 14:21:10,652 - rabbitpy.events - DEBUG - Waiting for 1 seconds on event: Socket Connected
2015-05-18 14:21:10,774 - rabbitpy.io - DEBUG - Connecting to ('188.165.36.67', 5672)
2015-05-18 14:21:10,797 - rabbitpy.base - DEBUG - IO setting state to 'Open'
2015-05-18 14:21:10,797 - rabbitpy.io - DEBUG - Socket connected
2015-05-18 14:21:10,798 - rabbitpy.io - DEBUG - Returning PollPoller
2015-05-18 14:21:10,816 - rabbitpy.connection - DEBUG - Adding channel 0 to io
2015-05-18 14:21:10,816 - rabbitpy.base - DEBUG - Channel0 setting state to 'Opening'
2015-05-18 14:21:10,817 - rabbitpy.base - DEBUG - Writing frame: ProtocolHeader
2015-05-18 14:21:10,844 - rabbitpy.channel0 - DEBUG - Received frame: 'Connection.Start'
2015-05-18 14:21:10,844 - rabbitpy.channel0 - DEBUG - Server information: 'Licensed under the MPL.  See http://www.rabbitmq.com/'
2015-05-18 14:21:10,845 - rabbitpy.channel0 - DEBUG - Server product: 'RabbitMQ'
2015-05-18 14:21:10,845 - rabbitpy.channel0 - DEBUG - Server copyright: 'Copyright (C) 2007-2014 GoPivotal, Inc.'
2015-05-18 14:21:10,845 - rabbitpy.channel0 - DEBUG - Server supports exchange_exchange_bindings: True
2015-05-18 14:21:10,845 - rabbitpy.channel0 - DEBUG - Server supports connection.blocked: True
2015-05-18 14:21:10,845 - rabbitpy.channel0 - DEBUG - Server supports authentication_failure_close: True
2015-05-18 14:21:10,845 - rabbitpy.channel0 - DEBUG - Server supports basic.nack: True
2015-05-18 14:21:10,845 - rabbitpy.channel0 - DEBUG - Server supports per_consumer_qos: True
2015-05-18 14:21:10,845 - rabbitpy.channel0 - DEBUG - Server supports consumer_priorities: True
2015-05-18 14:21:10,845 - rabbitpy.channel0 - DEBUG - Server supports consumer_cancel_notify: True
2015-05-18 14:21:10,845 - rabbitpy.channel0 - DEBUG - Server supports publisher_confirms: True
2015-05-18 14:21:10,845 - rabbitpy.channel0 - DEBUG - Server cluster_name: 'roger@rabbit.milanamos.net'
2015-05-18 14:21:10,845 - rabbitpy.channel0 - DEBUG - Server platform: 'Erlang/OTP'
2015-05-18 14:21:10,845 - rabbitpy.channel0 - DEBUG - Server version: '3.5.0'
2015-05-18 14:21:10,845 - rabbitpy.base - DEBUG - Writing frame: Connection.StartOk
2015-05-18 14:21:10,869 - rabbitpy.channel0 - DEBUG - Received frame: 'Connection.Tune'
2015-05-18 14:21:10,869 - rabbitpy.channel0 - DEBUG - Tuning, client: 300
2015-05-18 14:21:10,869 - rabbitpy.channel0 - DEBUG - Started a heartbeat timer that will fire in 450 sec
2015-05-18 14:21:10,870 - rabbitpy.base - DEBUG - Writing frame: Connection.TuneOk
2015-05-18 14:21:10,870 - rabbitpy.base - DEBUG - Writing frame: Connection.Open
2015-05-18 14:21:10,956 - rabbitpy.channel0 - DEBUG - Received frame: 'Connection.OpenOk'
2015-05-18 14:21:10,956 - rabbitpy.channel0 - DEBUG - Connection opened
2015-05-18 14:21:10,956 - rabbitpy.base - DEBUG - Channel0 setting state to 'Open'
2015-05-18 14:21:10,958 - store_stats - INFO - Connected to rabbitmq
2015-05-18 14:21:10,958 - rabbitpy.connection - DEBUG - Adding channel 1 to io
2015-05-18 14:21:10,958 - rabbitpy.base - DEBUG - Channel setting state to 'Opening'
2015-05-18 14:21:10,958 - rabbitpy.base - DEBUG - Writing frame: Channel.Open
2015-05-18 14:21:10,958 - rabbitpy.base - DEBUG - Waiting on <class 'pamqp.specification.OpenOk'> frame(s)
2015-05-18 14:21:10,990 - rabbitpy.channel - DEBUG - Checking for RPC request: <pamqp.specification.Channel.OpenOk object at 0x2ae3750>
2015-05-18 14:21:10,990 - rabbitpy.base - DEBUG - Channel setting state to 'Open'
2015-05-18 14:21:10,990 - rabbitpy.channel - DEBUG - Channel #1 open
2015-05-18 14:21:10,990 - store_stats - INFO - Channel opened
2015-05-18 14:21:10,991 - rabbitpy.base - DEBUG - Sending 'Basic.Consume'
2015-05-18 14:21:10,991 - rabbitpy.base - DEBUG - Writing frame: Basic.Consume
2015-05-18 14:21:10,991 - rabbitpy.base - DEBUG - Waiting on 'Basic.ConsumeOk' frame(s)
2015-05-18 14:21:11,054 - rabbitpy.channel - DEBUG - Checking for RPC request: <pamqp.specification.Basic.ConsumeOk object at 0x2ae37f8>
2015-05-18 14:21:11,054 - rabbitpy.base - DEBUG - Waiting on <class 'pamqp.specification.Deliver'> frame(s)
2015-05-18 14:26:10,896 - rabbitpy.channel0 - DEBUG - Received frame: 'Heartbeat'
2015-05-18 14:26:10,896 - rabbitpy.channel0 - DEBUG - Received Heartbeat, cancelling check timer
2015-05-18 14:26:10,896 - rabbitpy.base - DEBUG - Writing frame: Heartbeat
2015-05-18 14:26:10,896 - rabbitpy.channel0 - DEBUG - Started a heartbeat timer that will fire in 450 sec
2015-05-18 14:31:10,899 - rabbitpy.channel0 - DEBUG - Received frame: 'Heartbeat'
2015-05-18 14:31:10,899 - rabbitpy.channel0 - DEBUG - Received Heartbeat, cancelling check timer
2015-05-18 14:31:10,899 - rabbitpy.base - DEBUG - Writing frame: Heartbeat
2015-05-18 14:31:10,899 - rabbitpy.channel0 - DEBUG - Started a heartbeat timer that will fire in 450 sec
2015-05-18 14:35:49,881 - rabbitpy.channel - DEBUG - Checking for RPC request: <pamqp.specification.Basic.Deliver object at 0x2acedb8>
2015-05-18 14:35:49,881 - rabbitpy.channel - DEBUG - Waiting on content frames for Basic.Deliver: 1
2015-05-18 14:35:49,881 - rabbitpy.base - DEBUG - Waiting on 'ContentHeader' frame(s)
2015-05-18 14:35:49,881 - rabbitpy.channel - DEBUG - Checking for RPC request: <pamqp.header.ContentHeader object at 0x2ae7710>
2015-05-18 14:35:49,881 - rabbitpy.channel - DEBUG - Checking for RPC request: <pamqp.header.ContentHeader object at 0x2ae7710>
2015-05-18 14:35:49,881 - rabbitpy.base - DEBUG - Waiting on 'ContentBody' frame(s)
2015-05-18 14:35:49,881 - rabbitpy.channel - DEBUG - Checking for RPC request: <pamqp.body.ContentBody object at 0x2ae7790>
2015-05-18 14:35:49,881 - rabbitpy.channel - DEBUG - Checking for RPC request: <pamqp.body.ContentBody object at 0x2ae7790>
2015-05-18 14:35:49,881 - store_stats - INFO - Message: <rabbitpy.message.Message object at 0x2ae77d0>, statistics.market_analysis_charts.xxx.xxx
2015-05-18 14:35:49,978 - rabbitpy.base - DEBUG - Writing frame: Basic.Ack
[... a few other messages received ...]
2015-05-18 14:51:48,511 - rabbitpy.base - DEBUG - Waiting on <class 'pamqp.specification.Deliver'> frame(s)
2015-05-18 14:52:43,142 - rabbitpy.channel - DEBUG - Checking for RPC request: <pamqp.specification.Basic.Deliver object at 0x2acefc8>
2015-05-18 14:52:43,142 - rabbitpy.channel - DEBUG - Waiting on content frames for Basic.Deliver: 10
2015-05-18 14:52:43,142 - rabbitpy.base - DEBUG - Waiting on 'ContentHeader' frame(s)
2015-05-18 14:52:43,142 - rabbitpy.channel - DEBUG - Checking for RPC request: <pamqp.header.ContentHeader object at 0x2ae7910>
2015-05-18 14:52:43,142 - rabbitpy.channel - DEBUG - Checking for RPC request: <pamqp.header.ContentHeader object at 0x2ae7910>
2015-05-18 14:52:43,142 - rabbitpy.base - DEBUG - Waiting on 'ContentBody' frame(s)
2015-05-18 14:52:43,142 - rabbitpy.channel - DEBUG - Checking for RPC request: <pamqp.body.ContentBody object at 0x2ae7350>
2015-05-18 14:52:43,142 - rabbitpy.channel - DEBUG - Checking for RPC request: <pamqp.body.ContentBody object at 0x2ae7350>
2015-05-18 14:52:43,142 - store_stats - INFO - Message: <rabbitpy.message.Message object at 0x2ae7690>, statistics.market_analysis.xxx.xxx
2015-05-18 14:52:43,196 - rabbitpy.base - DEBUG - Writing frame: Basic.Ack
2015-05-18 14:52:43,196 - rabbitpy.base - DEBUG - Waiting on <class 'pamqp.specification.Deliver'> frame(s)
2015-05-18 14:53:40,904 - rabbitpy.channel0 - DEBUG - Checking for heartbeat, last: 900 sec ago, threshold: 900
2015-05-18 14:53:40,905 - rabbitpy.channel0 - ERROR - Have not received a heartbeat in 900 seconds
2015-05-18 14:53:41,212 - rabbitpy.base - DEBUG - Waiting for interrupt to clear
2015-05-18 14:56:10,909 - rabbitpy.channel0 - DEBUG - Received frame: 'Heartbeat'
2015-05-18 14:56:10,909 - rabbitpy.base - DEBUG - Writing frame: Heartbeat
2015-05-18 14:56:10,909 - rabbitpy.channel0 - DEBUG - Started a heartbeat timer that will fire in 450 sec
2015-05-18 15:01:10,910 - rabbitpy.channel0 - DEBUG - Received frame: 'Heartbeat'
2015-05-18 15:01:10,910 - rabbitpy.channel0 - DEBUG - Received Heartbeat, cancelling check timer
2015-05-18 15:01:10,910 - rabbitpy.base - DEBUG - Writing frame: Heartbeat
2015-05-18 15:01:10,910 - rabbitpy.channel0 - DEBUG - Started a heartbeat timer that will fire in 450 sec
2015-05-18 15:08:40,911 - rabbitpy.channel0 - DEBUG - Checking for heartbeat, last: 450 sec ago, threshold: 900
2015-05-18 15:08:40,911 - rabbitpy.channel0 - DEBUG - Started a heartbeat timer that will fire in 450 sec
2015-05-18 15:08:40,916 - rabbitpy.channel0 - DEBUG - Received frame: 'Heartbeat'
2015-05-18 15:08:40,916 - rabbitpy.channel0 - DEBUG - Received Heartbeat, cancelling check timer
2015-05-18 15:08:40,916 - rabbitpy.base - DEBUG - Writing frame: Heartbeat

```

From the rabbitmq console, I can see that the new messages (messages created after the ERROR at 14:53:40) are "unacked" and stay like this. The program is not aware of anything going wrong underneath :-(

**Comment by gmr** [comment]
This should be addressed in master, if you're able to test against that, please let me know your results, otherwise, 0.27.0 should drop real soon.

**Comment by rejoc** [comment]
I've installed current master version this morning. Everything is fine up to now but, I have no way to simulate a problem

**Comment by gmr** [comment]
Thanks for the update

**Comment by rejoc** [comment]
Any date about the release of a new version that integrate these changes ?

**Comment by ValentinH** [comment]
When 0.27 will be released?

**Comment by cmercier** [comment]
Hi, I would like to know when those changes will be released to avoid to install it from the source.

Thanks in advance

**Comment by rev138** [comment]
I am also looking for 2.7. The "real soon now" message was almost exactly a year ago.

**Comment by gmr** [comment]
This is addressed in 1.0

**Comment by ValentinH** [comment]
\o/
